File: 01-foundations/feature-engineering/code/python/feature_engineering_pipeline.py
"""
Feature Engineering Pipeline
=============================

Comprehensive feature engineering examples for machine learning.

Topics covered:
- Missing value imputation
- Categorical encoding
- Feature scaling and transformation
- Feature creation and interaction terms
- Dimensionality reduction
- Feature selection
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.feature_selection import SelectKBest, f_classif, mutual_info_classif
from sklearn.feature_selection import RFE
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from typing import List, Tuple
import warnings
warnings.filterwarnings('ignore')

sns.set_style("whitegrid")


class FeatureEngineeringPipeline:
    """
    Comprehensive feature engineering pipeline.

    Parameters
    ----------
    numerical_features : list
        Names of numerical features
    categorical_features : list
        Names of categorical features
    target : str
        Target variable name
    """

    def __init__(self, numerical_features: List[str],
                 categorical_features: List[str],
                 target: str = 'target'):
        self.numerical_features = numerical_features
        self.categorical_features = categorical_features
        self.target = target
        self.pipeline = None
        self.feature_names = None

    def create_basic_pipeline(self, strategy: str = 'robust') -> ColumnTransformer:
        """
        Create a basic preprocessing pipeline.

        Parameters
        ----------
        strategy : str
            Scaling strategy: 'standard', 'minmax', or 'robust'

        Returns
        -------
        ColumnTransformer
            Preprocessing pipeline
        """
        # Numerical pipeline
        if strategy == 'standard':
            scaler = StandardScaler()
        elif strategy == 'minmax':
            scaler = MinMaxScaler()
        else:
            scaler = RobustScaler()

        numerical_pipeline = Pipeline([
            ('imputer', SimpleImputer(strategy='median')),
            ('scaler', scaler)
        ])

        # Categorical pipeline
        categorical_pipeline = Pipeline([
            ('imputer', SimpleImputer(strategy='most_frequent')),
            ('encoder', OneHotEncoder(drop='first', sparse_output=False,
                                     handle_unknown='ignore'))
        ])

        # Combine pipelines
        preprocessor = ColumnTransformer(
            transformers=[
                ('num', numerical_pipeline, self.numerical_features),
                ('cat', categorical_pipeline, self.categorical_features)
            ],
            remainder='passthrough'
        )

        self.pipeline = preprocessor
        return preprocessor

    def create_polynomial_features(self, df: pd.DataFrame, degree: int = 2,
                                   interaction_only: bool = False) -> pd.DataFrame:
        """
        Create polynomial and interaction features.

        Parameters
        ----------
        df : pd.DataFrame
            Input dataframe
        degree : int
            Polynomial degree
        interaction_only : bool
            Only create interaction terms (no powers)

        Returns
        -------
        pd.DataFrame
            DataFrame with polynomial features
        """
        from sklearn.preprocessing import PolynomialFeatures

        poly = PolynomialFeatures(degree=degree,
                                 interaction_only=interaction_only,
                                 include_bias=False)

        num_data = df[self.numerical_features].values
        poly_features = poly.fit_transform(num_data)

        # Create feature names
        feature_names = poly.get_feature_names_out(self.numerical_features)

        poly_df = pd.DataFrame(poly_features, columns=feature_names,
                              index=df.index)

        # Combine with categorical features
        if self.categorical_features:
            result = pd.concat([poly_df, df[self.categorical_features]], axis=1)
        else:
            result = poly_df

        return result

    def create_domain_features(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Create domain-specific features (example for demo).

        Parameters
        ----------
        df : pd.DataFrame
            Input dataframe

        Returns
        -------
        pd.DataFrame
            DataFrame with new features
        """
        df_new = df.copy()

        # Example: ratio features
        if len(self.numerical_features) >= 2:
            f1, f2 = self.numerical_features[0], self.numerical_features[1]
            df_new[f'{f1}_to_{f2}_ratio'] = df_new[f1] / (df_new[f2] + 1e-8)

        # Example: binning
        if len(self.numerical_features) >= 1:
            f = self.numerical_features[0]
            df_new[f'{f}_binned'] = pd.cut(df_new[f], bins=5,
                                          labels=['very_low', 'low', 'medium',
                                                 'high', 'very_high'])

        return df_new

    def select_features_univariate(self, X: pd.DataFrame, y: pd.Series,
                                   k: int = 10) -> List[str]:
        """
        Select top k features using univariate statistical tests.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix
        y : pd.Series
            Target variable
        k : int
            Number of features to select

        Returns
        -------
        list
            Selected feature names
        """
        # Ensure numerical data only
        X_num = X.select_dtypes(include=[np.number])

        selector = SelectKBest(f_classif, k=min(k, X_num.shape[1]))
        selector.fit(X_num, y)

        # Get selected feature names
        selected_mask = selector.get_support()
        selected_features = X_num.columns[selected_mask].tolist()

        # Get scores
        scores = pd.DataFrame({
            'feature': X_num.columns,
            'score': selector.scores_
        }).sort_values('score', ascending=False)

        print("\nTop 10 features by F-score:")
        print(scores.head(10))

        return selected_features

    def select_features_rfe(self, X: pd.DataFrame, y: pd.Series,
                           n_features: int = 10) -> List[str]:
        """
        Recursive Feature Elimination using Random Forest.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix
        y : pd.Series
            Target variable
        n_features : int
            Number of features to select

        Returns
        -------
        list
            Selected feature names
        """
        X_num = X.select_dtypes(include=[np.number])

        estimator = RandomForestClassifier(n_estimators=100, random_state=42)
        selector = RFE(estimator, n_features_to_select=n_features, step=1)
        selector.fit(X_num, y)

        selected_mask = selector.get_support()
        selected_features = X_num.columns[selected_mask].tolist()

        # Show ranking
        ranking = pd.DataFrame({
            'feature': X_num.columns,
            'ranking': selector.ranking_
        }).sort_values('ranking')

        print("\nFeature rankings (1 = selected):")
        print(ranking.head(15))

        return selected_features

    def dimensionality_reduction_pca(self, X: pd.DataFrame,
                                    n_components: int = 2,
                                    plot: bool = True) -> Tuple[np.ndarray, PCA]:
        """
        Apply PCA for dimensionality reduction.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix
        n_components : int
            Number of principal components
        plot : bool
            Whether to plot explained variance

        Returns
        -------
        tuple
            (Transformed data, fitted PCA object)
        """
        X_num = X.select_dtypes(include=[np.number])

        # Standardize first
        scaler = StandardScaler()
        X_scaled = scaler.fit_transform(X_num)

        # Apply PCA
        pca = PCA(n_components=n_components)
        X_pca = pca.fit_transform(X_scaled)

        if plot:
            fig, axes = plt.subplots(1, 2, figsize=(14, 5))

            # Explained variance
            axes[0].bar(range(1, n_components + 1),
                       pca.explained_variance_ratio_)
            axes[0].set_xlabel('Principal Component')
            axes[0].set_ylabel('Explained Variance Ratio')
            axes[0].set_title('PCA Explained Variance')

            # Cumulative variance
            cumvar = np.cumsum(pca.explained_variance_ratio_)
            axes[1].plot(range(1, n_components + 1), cumvar, 'o-')
            axes[1].set_xlabel('Number of Components')
            axes[1].set_ylabel('Cumulative Explained Variance')
            axes[1].set_title('Cumulative Explained Variance')
            axes[1].axhline(y=0.95, color='r', linestyle='--',
                          label='95% variance')
            axes[1].legend()

            plt.tight_layout()
            plt.savefig('pca_analysis.png', dpi=150, bbox_inches='tight')
            plt.show()

        print(f"\nExplained variance ratio: {pca.explained_variance_ratio_}")
        print(f"Total variance explained: {pca.explained_variance_ratio_.sum():.2%}")

        return X_pca, pca


def demonstrate_missing_values():
    """Demonstrate different missing value imputation strategies."""
    print("=" * 70)
    print("Missing Value Imputation Strategies")
    print("=" * 70)

    np.random.seed(42)
    n = 1000

    # Create data with missing values
    df = pd.DataFrame({
        'age': np.random.normal(40, 15, n),
        'income': np.random.lognormal(10, 1, n),
        'score': np.random.normal(70, 15, n)
    })

    # Introduce missing values
    missing_mask = np.random.random(n) < 0.2
    df.loc[missing_mask, 'income'] = np.nan

    print(f"\nMissing values: {df.isnull().sum()['income']} / {n}")

    # Different imputation strategies
    strategies = {
        'Mean': SimpleImputer(strategy='mean'),
        'Median': SimpleImputer(strategy='median'),
        'KNN': KNNImputer(n_neighbors=5)
    }

    results = {}
    for name, imputer in strategies.items():
        df_imputed = df.copy()
        df_imputed[['income']] = imputer.fit_transform(df[['income']])
        results[name] = df_imputed['income']

    # Plot comparison
    fig, axes = plt.subplots(1, 4, figsize=(16, 4))

    # Original data (before introducing missing values)
    true_income = np.random.lognormal(10, 1, n)
    axes[0].hist(true_income[missing_mask], bins=30, alpha=0.7, color='gray')
    axes[0].set_title('True Values\n(before missing)')
    axes[0].set_xlabel('Income')

    for idx, (name, imputed) in enumerate(results.items(), 1):
        axes[idx].hist(imputed[missing_mask], bins=30, alpha=0.7)
        axes[idx].set_title(f'{name} Imputation')
        axes[idx].set_xlabel('Income')

    plt.tight_layout()
    plt.savefig('imputation_comparison.png', dpi=150, bbox_inches='tight')
    plt.show()


def demonstrate_encoding():
    """Demonstrate categorical encoding strategies."""
    print("\n" + "=" * 70)
    print("Categorical Encoding Strategies")
    print("=" * 70)

    # Sample data
    df = pd.DataFrame({
        'city': ['NYC', 'LA', 'Chicago', 'NYC', 'LA'] * 20,
        'education': ['HS', 'BS', 'MS', 'PhD', 'BS'] * 20,
        'salary': np.random.normal(70000, 20000, 100)
    })

    print("\nOriginal data sample:")
    print(df.head())

    # 1. Label Encoding
    le = LabelEncoder()
    df['city_label'] = le.fit_transform(df['city'])

    # 2. One-Hot Encoding
    df_onehot = pd.get_dummies(df, columns=['city', 'education'],
                               prefix=['city', 'edu'], drop_first=True)

    # 3. Ordinal Encoding (for education)
    education_order = ['HS', 'BS', 'MS', 'PhD']
    ordinal_encoder = OrdinalEncoder(categories=[education_order])
    df['education_ordinal'] = ordinal_encoder.fit_transform(df[['education']])

    # 4. Target Encoding (mean encoding)
    city_means = df.groupby('city')['salary'].mean()
    df['city_target_encoded'] = df['city'].map(city_means)

    print("\nEncoding results:")
    print(f"Label encoding: {df[['city', 'city_label']].head()}")
    print(f"\nOne-hot encoding columns: {[c for c in df_onehot.columns if 'city_' in c or 'edu_' in c]}")
    print(f"\nOrdinal encoding: {df[['education', 'education_ordinal']].head()}")
    print(f"\nTarget encoding: {df[['city', 'city_target_encoded']].head()}")


def demonstrate_complete_pipeline():
    """Demonstrate a complete feature engineering pipeline."""
    print("\n" + "=" * 70)
    print("Complete Feature Engineering Pipeline")
    print("=" * 70)

    # Load or create sample data
    from sklearn.datasets import make_classification

    X, y = make_classification(n_samples=1000, n_features=20, n_informative=10,
                               n_redundant=5, n_classes=2, random_state=42)

    # Create DataFrame
    feature_names = [f'feature_{i}' for i in range(20)]
    df = pd.DataFrame(X, columns=feature_names)
    df['target'] = y

    # Add some categorical features
    df['category_A'] = np.random.choice(['cat1', 'cat2', 'cat3'], 1000)
    df['category_B'] = np.random.choice(['low', 'medium', 'high'], 1000)

    # Introduce missing values
    for col in feature_names[:5]:
        mask = np.random.random(1000) < 0.1
        df.loc[mask, col] = np.nan

    print(f"\nDataset shape: {df.shape}")
    print(f"Missing values:\n{df.isnull().sum()[df.isnull().sum() > 0]}")

    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        df.drop('target', axis=1), df['target'],
        test_size=0.2, random_state=42
    )

    # Create pipeline
    numerical_features = feature_names
    categorical_features = ['category_A', 'category_B']

    fe = FeatureEngineeringPipeline(numerical_features, categorical_features)

    # Build and apply pipeline
    preprocessor = fe.create_basic_pipeline(strategy='robust')
    X_train_transformed = preprocessor.fit_transform(X_train)
    X_test_transformed = preprocessor.transform(X_test)

    print(f"\nTransformed data shape: {X_train_transformed.shape}")

    # Feature selection
    X_train_df = pd.DataFrame(X_train_transformed,
                             columns=range(X_train_transformed.shape[1]))
    selected_features = fe.select_features_univariate(X_train_df, y_train, k=10)

    # Train model
    clf = RandomForestClassifier(n_estimators=100, random_state=42)
    clf.fit(X_train_transformed, y_train)

    # Feature importance
    importances = pd.DataFrame({
        'feature': range(len(clf.feature_importances_)),
        'importance': clf.feature_importances_
    }).sort_values('importance', ascending=False)

    plt.figure(figsize=(10, 6))
    plt.bar(range(15), importances.head(15)['importance'])
    plt.xlabel('Feature Index')
    plt.ylabel('Importance')
    plt.title('Top 15 Feature Importances')
    plt.tight_layout()
    plt.savefig('feature_importance.png', dpi=150, bbox_inches='tight')
    plt.show()

    print("\nTop 10 most important features:")
    print(importances.head(10))

    # Model performance
    from sklearn.metrics import classification_report
    y_pred = clf.predict(X_test_transformed)
    print("\nModel Performance:")
    print(classification_report(y_test, y_pred))


if __name__ == "__main__":
    demonstrate_missing_values()
    demonstrate_encoding()
    demonstrate_complete_pipeline()

    print("\nAll feature engineering examples completed!")

File: 01-foundations/feature-engineering/code/r/feature_engineering_pipeline.R
# Feature Engineering Pipeline in R
# ==================================
#
# Comprehensive feature engineering examples for machine learning

# Load required packages
suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
  library(caret)        # For preprocessing
  library(recipes)      # For feature engineering pipelines
  library(ggplot2)
  library(corrplot)
  library(mice)         # For missing data imputation
  library(randomForest)
})

# ============================================
# Missing Value Imputation
# ============================================

demonstrate_imputation <- function() {
  cat("=" * 70, "\n")
  cat("Missing Value Imputation Strategies\n")
  cat("=" * 70, "\n\n")

  set.seed(42)
  n <- 1000

  # Create data with missing values
  df <- data.frame(
    age = rnorm(n, 40, 15),
    income = rlnorm(n, 10, 1),
    score = rnorm(n, 70, 15)
  )

  # Introduce missing values (20% missing)
  missing_indices <- sample(1:n, 0.2 * n)
  df$income[missing_indices] <- NA

  cat(sprintf("Missing values: %d / %d (%.1f%%)\n",
             sum(is.na(df$income)), n, 100 * mean(is.na(df$income))))

  # Method 1: Mean imputation
  df$income_mean <- df$income
  df$income_mean[is.na(df$income_mean)] <- mean(df$income, na.rm = TRUE)

  # Method 2: Median imputation
  df$income_median <- df$income
  df$income_median[is.na(df$income_median)] <- median(df$income, na.rm = TRUE)

  # Method 3: KNN imputation (using caret)
  preproc <- preProcess(df[, "income", drop = FALSE],
                       method = "knnImpute", k = 5)
  df$income_knn <- predict(preproc, df[, "income", drop = FALSE])[[1]]

  # Method 4: MICE (Multiple Imputation)
  mice_model <- mice(df[, c("age", "income", "score")],
                    m = 1, method = "pmm", seed = 42, printFlag = FALSE)
  df$income_mice <- complete(mice_model)$income

  # Compare methods
  cat("\nImputation method comparison:\n")
  cat(sprintf("Mean:   %.2f\n", mean(df$income_mean)))
  cat(sprintf("Median: %.2f\n", mean(df$income_median)))
  cat(sprintf("KNN:    %.2f\n", mean(df$income_knn)))
  cat(sprintf("MICE:   %.2f\n", mean(df$income_mice)))
  cat(sprintf("Original (w/o NA): %.2f\n", mean(df$income, na.rm = TRUE)))

  invisible(df)
}

# ============================================
# Categorical Encoding
# ============================================

demonstrate_encoding <- function() {
  cat("\n", "=" * 70, "\n")
  cat("Categorical Encoding Strategies\n")
  cat("=" * 70, "\n\n")

  # Sample data
  df <- data.frame(
    city = rep(c("NYC", "LA", "Chicago"), length.out = 100),
    education = rep(c("HS", "BS", "MS", "PhD"), length.out = 100),
    salary = rnorm(100, 70000, 20000),
    stringsAsFactors = FALSE
  )

  cat("Original data sample:\n")
  print(head(df))

  # 1. Label Encoding (factor levels)
  df$city_label <- as.numeric(factor(df$city))

  # 2. One-Hot Encoding
  dummy_vars <- dummyVars(~ city + education, data = df, fullRank = TRUE)
  df_onehot <- predict(dummy_vars, df)
  df_encoded <- cbind(df, df_onehot)

  # 3. Ordinal Encoding (for education)
  education_order <- c("HS", "BS", "MS", "PhD")
  df$education_ordinal <- match(df$education, education_order)

  # 4. Target Encoding (mean encoding)
  city_means <- df %>%
    group_by(city) %>%
    summarise(mean_salary = mean(salary), .groups = "drop")

  df <- df %>%
    left_join(city_means, by = "city") %>%
    rename(city_target_encoded = mean_salary)

  cat("\n\nEncoding results:\n")
  cat("\nLabel encoding (first 5 rows):\n")
  print(head(df[, c("city", "city_label")], 5))

  cat("\nOrdinal encoding (first 5 rows):\n")
  print(head(df[, c("education", "education_ordinal")], 5))

  cat("\nTarget encoding (first 5 rows):\n")
  print(head(df[, c("city", "city_target_encoded")], 5))

  invisible(df)
}

# ============================================
# Feature Scaling
# ============================================

demonstrate_scaling <- function() {
  cat("\n", "=" * 70, "\n")
  cat("Feature Scaling Strategies\n")
  cat("=" * 70, "\n\n")

  set.seed(42)
  df <- data.frame(
    feature1 = rnorm(100, 50, 10),
    feature2 = rnorm(100, 1000, 500),
    feature3 = rexp(100, 0.1)
  )

  cat("Original data statistics:\n")
  print(summary(df))

  # 1. Standardization (Z-score normalization)
  df_standard <- as.data.frame(scale(df))
  names(df_standard) <- paste0(names(df), "_standard")

  # 2. Min-Max scaling
  preproc_minmax <- preProcess(df, method = "range")
  df_minmax <- predict(preproc_minmax, df)
  names(df_minmax) <- paste0(names(df), "_minmax")

  # 3. Robust scaling (median and IQR)
  robust_scale <- function(x) {
    (x - median(x)) / IQR(x)
  }
  df_robust <- as.data.frame(lapply(df, robust_scale))
  names(df_robust) <- paste0(names(df), "_robust")

  cat("\n\nStandardized data statistics:\n")
  print(summary(df_standard))

  invisible(list(standard = df_standard, minmax = df_minmax, robust = df_robust))
}

# ============================================
# Feature Creation with recipes
# ============================================

create_feature_pipeline <- function() {
  cat("\n", "=" * 70, "\n")
  cat("Feature Engineering Pipeline with recipes\n")
  cat("=" * 70, "\n\n")

  # Sample data
  set.seed(123)
  n <- 500

  df <- data.frame(
    age = rnorm(n, 40, 15),
    income = rlnorm(n, 10, 1),
    education = sample(c("HS", "BS", "MS", "PhD"), n, replace = TRUE),
    city = sample(c("NYC", "LA", "Chicago", "Houston"), n, replace = TRUE),
    outcome = rbinom(n, 1, 0.5)
  )

  # Introduce missing values
  df$income[sample(1:n, 0.1 * n)] <- NA

  # Create recipe
  rec <- recipe(outcome ~ ., data = df) %>%
    # Impute missing values
    step_impute_median(all_numeric(), -all_outcomes()) %>%
    step_impute_mode(all_nominal(), -all_outcomes()) %>%

    # Create interaction terms
    step_interact(terms = ~ age:income) %>%

    # Create polynomial features
    step_poly(age, degree = 2) %>%

    # Log transform
    step_log(income, offset = 1) %>%

    # One-hot encoding
    step_dummy(all_nominal(), -all_outcomes(), one_hot = FALSE) %>%

    # Standardize
    step_normalize(all_numeric(), -all_outcomes()) %>%

    # Remove zero-variance features
    step_zv(all_predictors())

  # Prepare recipe
  rec_prepped <- prep(rec, training = df)

  # Apply to data
  df_processed <- bake(rec_prepped, new_data = df)

  cat("Original data shape: ", dim(df), "\n")
  cat("Processed data shape:", dim(df_processed), "\n")
  cat("\nProcessed features:\n")
  print(names(df_processed))

  invisible(list(recipe = rec_prepped, data = df_processed))
}

# ============================================
# Feature Selection
# ============================================

demonstrate_feature_selection <- function() {
  cat("\n", "=" * 70, "\n")
  cat("Feature Selection Methods\n")
  cat("=" * 70, "\n\n")

  # Load sample data
  data(iris)
  df <- iris
  df$Species <- as.factor(df$Species)

  # Method 1: Correlation-based filtering
  cor_matrix <- cor(df[, 1:4])
  cat("Correlation matrix:\n")
  print(round(cor_matrix, 2))

  # Find highly correlated features (|r| > 0.8)
  high_cor <- findCorrelation(cor_matrix, cutoff = 0.8)
  if (length(high_cor) > 0) {
    cat(sprintf("\nHighly correlated features to remove: %s\n",
               paste(names(df)[high_cor], collapse = ", ")))
  }

  # Method 2: Random Forest importance
  set.seed(42)
  rf_model <- randomForest(Species ~ ., data = df, importance = TRUE, ntree = 500)

  importance_df <- data.frame(
    Feature = rownames(importance(rf_model)),
    Importance = importance(rf_model)[, "MeanDecreaseGini"]
  )
  importance_df <- importance_df[order(-importance_df$Importance), ]

  cat("\nRandom Forest Feature Importance:\n")
  print(importance_df)

  # Method 3: Recursive Feature Elimination (RFE)
  control <- rfeControl(functions = rfFuncs, method = "cv", number = 5)
  rfe_results <- rfe(df[, 1:4], df$Species,
                    sizes = c(1:4),
                    rfeControl = control)

  cat("\n\nRFE Results:\n")
  print(rfe_results)
  cat("\nOptimal features:", predictors(rfe_results), "\n")

  # Plot importance
  varImpPlot(rf_model, main = "Feature Importance from Random Forest")

  invisible(list(rf = rf_model, rfe = rfe_results))
}

# ============================================
# Dimensionality Reduction (PCA)
# ============================================

demonstrate_pca <- function() {
  cat("\n", "=" * 70, "\n")
  cat("Principal Component Analysis (PCA)\n")
  cat("=" * 70, "\n\n")

  # Load data
  data(iris)
  df_numeric <- iris[, 1:4]

  # Standardize data
  df_scaled <- scale(df_numeric)

  # Apply PCA
  pca_result <- prcomp(df_scaled, center = FALSE, scale. = FALSE)

  # Summary
  cat("PCA Summary:\n")
  print(summary(pca_result))

  # Variance explained
  var_explained <- pca_result$sdev^2 / sum(pca_result$sdev^2)
  cumvar <- cumsum(var_explained)

  cat("\n\nVariance explained by each PC:\n")
  for (i in 1:4) {
    cat(sprintf("PC%d: %.2f%% (cumulative: %.2f%%)\n",
               i, 100 * var_explained[i], 100 * cumvar[i]))
  }

  # Scree plot
  par(mfrow = c(1, 2))

  # Variance explained
  barplot(var_explained, names.arg = paste0("PC", 1:4),
         main = "Variance Explained by Each PC",
         ylab = "Proportion of Variance",
         col = "steelblue")

  # Cumulative variance
  plot(1:4, cumvar, type = "b", pch = 19,
      main = "Cumulative Variance Explained",
      xlab = "Number of Components",
      ylab = "Cumulative Proportion",
      ylim = c(0, 1))
  abline(h = 0.95, col = "red", lty = 2)
  legend("bottomright", legend = "95% threshold", col = "red", lty = 2)

  par(mfrow = c(1, 1))

  # Biplot
  biplot(pca_result, scale = 0, cex = 0.7,
        main = "PCA Biplot")

  invisible(pca_result)
}

# ============================================
# Complete End-to-End Example
# ============================================

complete_pipeline_example <- function() {
  cat("\n", "=" * 70, "\n")
  cat("Complete Feature Engineering Pipeline\n")
  cat("=" * 70, "\n\n")

  # Create synthetic data
  set.seed(456)
  n <- 1000

  df <- data.frame(
    age = rnorm(n, 40, 15),
    income = rlnorm(n, 10, 1),
    education = sample(c("HS", "BS", "MS", "PhD"), n, replace = TRUE,
                      prob = c(0.3, 0.4, 0.2, 0.1)),
    city = sample(c("NYC", "LA", "Chicago", "Houston"), n, replace = TRUE),
    experience = rnorm(n, 10, 5),
    outcome = rbinom(n, 1, 0.5)
  )

  # Introduce missing values
  df$income[sample(1:n, 0.15 * n)] <- NA
  df$experience[sample(1:n, 0.1 * n)] <- NA

  cat(sprintf("Dataset: %d rows, %d columns\n", nrow(df), ncol(df)))
  cat("\nMissing values:\n")
  print(colSums(is.na(df)))

  # Build comprehensive recipe
  rec <- recipe(outcome ~ ., data = df) %>%
    # Missing value imputation
    step_impute_median(all_numeric_predictors()) %>%
    step_impute_mode(all_nominal_predictors()) %>%

    # Feature engineering
    step_interact(terms = ~ age:experience) %>%
    step_poly(age, degree = 2, options = list(raw = TRUE)) %>%
    step_log(income, offset = 1, base = 10) %>%

    # Binning
    step_discretize(experience, num_breaks = 4) %>%

    # Encoding
    step_dummy(all_nominal_predictors(), one_hot = FALSE) %>%

    # Scaling
    step_normalize(all_numeric_predictors()) %>%

    # Remove near-zero variance features
    step_nzv(all_predictors()) %>%

    # Remove highly correlated features
    step_corr(all_numeric_predictors(), threshold = 0.9)

  # Prepare and apply recipe
  rec_prepped <- prep(rec, training = df)
  df_processed <- bake(rec_prepped, new_data = df)

  cat("\n\nProcessed data:\n")
  cat(sprintf("Shape: %d rows, %d columns\n",
             nrow(df_processed), ncol(df_processed)))
  cat("\nFeature names:\n")
  print(names(df_processed))

  # Train model
  library(randomForest)
  rf_model <- randomForest(outcome ~ ., data = df_processed,
                          ntree = 500, importance = TRUE)

  cat("\n\nRandom Forest Model Performance:\n")
  print(rf_model)

  # Feature importance
  imp <- importance(rf_model)
  imp_sorted <- imp[order(-imp[, "MeanDecreaseGini"]), , drop = FALSE]

  cat("\nTop 10 Most Important Features:\n")
  print(head(imp_sorted, 10))

  invisible(list(recipe = rec_prepped, model = rf_model, data = df_processed))
}

# ============================================
# Run all examples
# ============================================

if (sys.nframe() == 0) {
  cat("\n")
  cat("=" * 70, "\n")
  cat("Feature Engineering Pipeline Examples in R\n")
  cat("=" * 70, "\n\n")

  demonstrate_imputation()
  demonstrate_encoding()
  demonstrate_scaling()
  create_feature_pipeline()
  demonstrate_feature_selection()
  demonstrate_pca()
  complete_pipeline_example()

  cat("\n\nAll feature engineering examples completed!\n")
}
